### ISSUE TITLE
Replace gl/glext with dglOpengl or GLAD

### ISSUE THREAD

**Issue by Kagamma** [castle-engine/castle-engine#334]
Currently CGE is relying on `glext` for desktop's modern OpenGL. The problem is that `glext` only support OpenGL up to 4.0 at the moment, mean we miss a lot of nice stuff (in particular Compute Shader in OpenGL 4.3).

Maybe we should use Sascha Willems's `dglOpengl` header (https://github.com/SaschaWillems/dglOpenGL) instead? `dglOpenGL` supports latest OpenGL version.

**Comment by Kagamma** [comment]
Another solution is to maintain our own glext copy and add missing functions & constants require for OpenGL > 4.0

**Comment by michaliskambi** [comment]
My first thought would be that we should submit a patch to FPC, to update their glext unit. From what I understand, they were made by Tom Nuydens (`tom@delphi3d.net` is mentioned in credits) though delphi3d.net seems no longer available.

That said, in parallel @and3md is working on Delphi port. And he's using (for now) Delphi-provided headers, `OpenGL` and `OpenGLExt` from what I understand. But we are not yet sure whether it will be the final state. And what will be used for OpenGLES on Delphi.

Maybe it will be better to use something like https://github.com/SaschaWillems/dglOpenGL if it is 1. maintained, 2. works with both FPC and Delphi (and it seems the answer is 2x "yes").

So... 

- no decision yet, 

- https://github.com/SaschaWillems/dglOpenGL seems nice, let's wait for Delphi port getting more advanced -- it may well force us to use dglOpenGL, if it covers both FPC and Delphi.

- regardless of it, I'd say to update FPC glext unit.

**Comment by michaliskambi** [comment]
Updating this with label "backlog - undecided". We may get back to this if some compiler/platforms forces us to this, in particular likely Delphi on non-Windows will force us to this.

**Comment by michaliskambi** [comment]
Update about this: it's no longer "backlog undecided" :) It's decided, we'll do it, that is: 

We will switch to use 

1. Either dglOpenGL (from https://github.com/SaschaWillems/dglOpenGL ) for Castle Game Engine. For all platforms, for both FPC and Delphi.  ( This applies to desktop OpenGL usage. For OpenGLES, we'll continue using CastleGLES. )

2. Or to use GLAD, https://github.com/Dav1dde/glad . For FPC and Delphi. For OpenGL and OpenGLES.

During my recent rendering code refactor I found more and more reasons to do it:

- I had to deal with some missing bits in FPC GL / GLExt.

- I had to avoid GL / GLExt issues with reading extensions (it doesn't support "core" OpenGL 3.0 way of `glGetStringi(GL_EXTENSIONS, index)`; it is possible to avoid using it, our `src/base_rendering/castleglutils_features.inc` does the dirty job, but it's a pain -- one has to be super-careful about it). I confirmed that `dglOpenGL.pas` has this fixed, it queries `glGetString(GL_VERSION)` and if it detects version >= 3 it is using the modern `glGetStringi(GL_EXTENSIONS, index)` automatically. Also GLAD generated code does it correctly.

- We have a few things I'd like to get rid of: `castleinternalglutils_delphi_wgl.inc`,  `castleinternalglutils_wgl.inc` . I confirmed `dglOpenGL.pas`  has all these WGL extensions.

- Delphi's `OpenGL, OpenGLext,` are tied to Windows. Delphi also has `Androidapi.JNI.OpenGL`, but that's an incompatible beast. Delphi also has `Macapi.OpenGL` but that's only with features up to OpenGL 2.1 (and there's no `Macapi.OpenGLext`). There's nothing for other platforms, like Delphi on Linux. 

    And since we now know that we'll add more platforms for Delphi, thanks to good response from Delphi users, we know we should deal with it sooner rather than later :)

My only remaining uncertainly here is:

Do we use dglOpenGL or GLAD?

1. https://github.com/SaschaWillems/dglOpenGL 
    - Supports OpenGL 4.6. No OpenGLES support.
    - It wasn't updated since 2017, well, but then maybe there was no need (OpenGL 4.6 is latest, and it was in 2017).
    - I see there is later movement on https://delphigl.com/ and forums https://delphigl.com/forum/index.php have posts even from "Feb 15, 2023". Unfortunately it's all German :), but at least we know there's a community there. 
    - It seems popular (e.g. Vampyre is using it for examples too.)

2. https://glad.dav1d.de/ . It is enticing, 
    -  generating code that looks good for both FPC and Delphi too, 
    - generating code for both OpenGL and OpenGLES, 
    - it uses specifications (XML files in https://github.com/Dav1dde/glad/tree/glad2/glad/files ) that are maintained by Khronos. So if we use this, we become super-independent and we don't need anyone to manually update OpenGL/OpenGLES Pascal headers in the future. We only need to make sure GLAD maintains templates for Pascal.
    - GLAD 2 ( https://gen.glad.sh/ ) doesn't support Pascal for now, unfortunately . But maybe we could add it?
    - Bonus: GLAD 2 supports also Vulkan.

I guess I'm looking for volunteers who'd like to play with GLAD, and in particular port GLAD 2 to Pascal. The directory https://github.com/Dav1dde/glad/tree/glad2/glad/generator contains now only C/C++ and Rust, but I'm guessing the author is open to PRs adding more languages. For comparison, GLAD 1 templates are in https://github.com/Dav1dde/glad/tree/master/glad/lang .

**Comment by michaliskambi** [comment]
This is in progress: Delphi+Linux branch (mentioned on news today, https://castle-engine.io/wp/2023/09/23/using-delphi-welcome-page-to-show-your-readme-md-also-docs-updates-also-coming-soon-new-navigation-components-and-delphi-linux/ ) brings dglOpenGL header. *For now* on that branch we only use it for Delphi on non-Windows, but the goal is to use it for *all platforms and compilers* for OpenGL. So we'll use just 2 headers: dglOpenGL (for OpenGL), and CastleGLES (for OpenGLES) with all compilers and platforms.

**Comment by michaliskambi** [comment]
P.S. In relation to my previous hesitation (dglOpenGL vs GLAD vs GLAD 2), it seems I have made a choice: dglOpenGL. Why?

- All in all, the choice fell on dglOpenGL as it seemed more widely used, and it's already available (unlike GLAD 2). 
- I knew dglOpenGL code better, and I was already sure it does some things correctly (e.g. using `glGetStringi(GL_EXTENSIONS, index)`)
- I was able to "just use it" in CGE with just a few simple code changes, so I'm happy.
- As Khronos officially said that OpenGL will not be updated with new features after 4.6 ( https://en.wikipedia.org/wiki/OpenGL ), and dglOpenGL supports 4.6 already, so we don't really care about "who will update this to newer spec versions" much.
- I am still very interested in getting GLAD 2 ported to Pascal and get Vulkan headers this way.
- Decision is not set in stone. If a good reason will come, maybe we will switch to GLAD (but there doesn't seem to be a convincing reason for it now). If GLAD 2 will be ported to Pascal, I will also look at it with interest (both for Vulkan and GL).
